### PR TITLE
[examples] Add trainer support for question-answering

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds trainer support for question-answering task. Regarding issue #4784

**TODOs**
- [ ]  Add automatic data loading. Right now it requires the user to specify data directory. Decided not to use `tfds` because I think it will be soon replaced by `nlp` here
- [ ] Add evaluation
- [ ] Test all models.

@julien-c  @patrickvonplaten

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hi @patil-suraj,

I think @julien-c can answer questions regarding the Trainer better :-)

File: examples/question-answering/run_squad_trainer.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+# coding=utf-8
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply examples/question-answering/run_squad_trainer.py:1]: Hi @daniel-shan, run_tf_squad.py is for Tensorflow, this PR adds Trainer support for fine-tuning pytorch models  for QA

[the remark this answers is not in the document]
Comment (daniel-shan) [reply examples/question-answering/run_squad_trainer.py:1]: Oops sorry didn't see that, thank you!

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -142,3 +142,21 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
...
+    """
+
+    def collate_batch(self, batch: List[Dict[str, torch.Tensor]]) -> Dict[str, torch.Tensor]:
+        keys = batch[0].keys()
+        inputs = {}
+
+        for key in keys:
+            inputs[key] = torch.stack([example[key] for example in batch])
+
+        return inputs
```
Comment (julien-c) [inline src/transformers/data/data_collator.py:162]: pretty clean implementation of data collator. cc @sgugger

**Review by julien-c — APPROVED** [review]
Looks great! 

My main question is, have you reproduced the training results that are documented in https://github.com/huggingface/transformers/blob/master/examples/question-answering/README.md ?

If you have, can you link to a webpage on your favorite experiment tracking service (weights and biases, cc @borisdayma:), Tensorboard.dev, etc) for reference?

Also you can just replace run_squad.py instead of creating a new file, but we can do this at the end/just before merging.

Finally, you can add a checkmark to the [Big Table Of Tasks](https://github.com/huggingface/transformers/tree/master/examples#the-big-table-of-tasks)

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/data/data_collator.py:162]: Yes, quick question though, why does the default not work here?

Comment (julien-c) [reply src/transformers/data/data_collator.py:162]: (I think it's ~identical to the default one so maybe we can merge them and just use the default here)

File: src/transformers/data/datasets/squad.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+class Split(Enum):
+    train = "train"
+    dev = "dev"
+
+
+class SquadDataset(Dataset):
+    """
+    This will be superseded by a framework-agnostic approach
+    soon.
```
Comment (julien-c) [inline src/transformers/data/datasets/squad.py:89]: Just adding a note/clarification that this will be replaced by the https://github.com/huggingface/nlp library soon-ish, cc @thomwolf

**Comment by borisdayma** [comment]
Just in case you wanted to use Weights & Biases, you should just have to do a `pip install wandb` and it should automatically track everything.

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (patil-suraj) [reply src/transformers/data/data_collator.py:162]: default collator needs `List[InputDataClass]` but I'm returning List[Dict`[str, torch.Tensor]]` so I was not able to use the default one

**Comment by patil-suraj** [comment]
>My main question is, have you reproduced the training results that are documented in https://github.com/huggingface/transformers/blob/master/examples/question-answering/README.md ?

I didn't train `bert-base` (just trained for 1 epoch to see if the implementation was working) but instead I used it to train `electra-base`  and it gave better results than mentioned in the paper

In the paper the authors mentioned that electra-base achieves 84.5 EM and 90.8 F1. I was able to achieve 85.05 EM and 91.60 F1. Sadly didn't use wandb, you can find the colab [here](https://colab.research.google.com/drive/11yo-LaFsgggwmDSy2P8zD3tzf5cCb-DU?usp=sharing)

It uses the same code, just copy pasted in colab. But if required I can try to reproduce the documented results.

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/data/data_collator.py:162]: InputDataClass is just a type alias for anything.

File: src/transformers/data/datasets/squad.py (hunk 3 above)
Comment (patil-suraj) [reply src/transformers/data/datasets/squad.py:89]: Yes, that will be actually better than this ;)

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (patil-suraj) [reply src/transformers/data/data_collator.py:162]: >InputDataClass is just a type alias for anything.

Yes, but I think it assumes that it will be class as it uses `getattr` and `vars`. How should I proceed ?

Comment (julien-c) [reply src/transformers/data/data_collator.py:162]: maybe we can change the default implem to accept more input types

**Comment by julien-c** [comment]
> I didn't train `bert-base` (just trained for 1 epoch to see if the implementation was working) 

I can do it tomorrow morning, I currently have a V100 on hand:)

**Comment by borisdayma** [comment]
Just a note that I tried `python run_squad_trainer.py --model_name_or_path bert-base-uncased --model_type bert --data_dir squad --output_dir /tmp/debug_squad/ --overwrite_output_dir --do_train --do_eval --evaluate_during_training --logging_steps 100`.

For some reason I don't get any evaluation metric during training (I was expecting `loss` or `eval_loss`).

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (julien-c) [reply src/transformers/data/data_collator.py:162]: Let me know if you want to do it @sgugger or @patil-suraj

Comment (patil-suraj) [reply src/transformers/data/data_collator.py:162]: @julien-c can we just add a simple type check in default collator, i.e if the input is `dict` we can call `example.get` instead of `getattr`, or maybe convert the `dict` to a simple `class` ?

**Comment by patil-suraj** [comment]
> Just in case you wanted to use Weights & Biases, you should just have to do a `pip install wandb` and it should automatically track everything.

@borisdayma yes, there are no start and end positions in eval dataset which is why eval loss is not calculated. I will add that. Were you able to see training loss ?
Thanks !

File: src/transformers/data/data_collator.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/data/data_collator.py:162]: I'll add this this morning with the backward compatibility to the old `DataCollator` style.

**Comment by julien-c** [comment]
> @borisdayma yes, there are no start and end positions in eval dataset which is why eval loss is not calculated. I will add that. Were you able to see training loss ?

Hmm, I'm pretty sure the dev-v1.1.json file has the same labels as the training one (start positions). Otherwise we wouldn't have any eval results at all in the readme. No?

**Comment by borisdayma** [comment]
> @borisdayma yes, there are no start and end positions in eval dataset which is why eval loss is not calculated. I will add that. Were you able to see training loss ?

Yes, training loss was logged.

**Comment by LysandreJik** [comment]
@julien-c In the two `TensorDatasets` created (one for training and one for evaluation), only the training has the correct `start_position` and `end_position`.

I believe this is because while the training dataset only has one possible answer per question, the dev and validation datasets both have multiple answers per question (usually different-lengths spans).

**Comment by julien-c** [comment]
@LysandreJik So I guess we should update the eval dataset to pick one start_position (or the most frequent one) – how do people do it usually with SQuAD eval, do you know @thomwolf?

Maybe this can be done in a second PR though. Everyone ok with merging this (renaming `run_squad_trainer.py` to `run_squad.py`)?

**Review by LysandreJik — APPROVED** [review]
Yes, sounds good!

**Comment by sgugger** [comment]
@patil-suraj Can you resolve the conflicts and switch to the new `default_data_collator` now that it should work for your dict inputs?
I can take over if you don't have time, but this is the only thing standing in the way of merging this PR.

**Comment by patil-suraj** [comment]
@sgugger Yes, I'll  switch to the new data collator.

**Comment by patil-suraj** [comment]
Hi @sgugger, you can take this over, I'm running short on time ;(

**Comment by patil-suraj** [comment]
Thanks @sgugger :)

**Comment by julien-c** [comment]
@sgugger can you please rename `run_squad_trainer.py` to `run_squad.py`? see also #5547
